src/Http/Http.Api/src/MapActionExpressionTreeBuilder.cs

```diff
@@ -0,0 +1,385 @@
+                                       TargetArg,
+                                       HttpContextParameter);
+                    }
+                }
+                else
+                {
+                    // TODO: Handle custom awaitables
+                    throw new NotSupportedException($"Unsupported return type: {method.ReturnType}");
+                }
+            }
+            else if (typeof(IResult).IsAssignableFrom(method.ReturnType))
+            {
+                body = Expression.Call(methodCall, ResultWriteResponseAsync, HttpContextParameter);
+            }
+            else if (method.ReturnType == typeof(string))
+            {
+                body = Expression.Call(StringResultWriteResponseAsync, HttpResponseExpr, methodCall, Expression.Constant(CancellationToken.None));
+            }
+            else
+            {
+                body = Expression.Call(JsonResultWriteResponseAsync, HttpResponseExpr, methodCall, Expression.Constant(CancellationToken.None));
+            }
+
+            Func<object?, HttpContext, ValueTask>? requestDelegate = null;
+
+            if (needBody)
+            {
+                // We need to generate the code for reading from the body before calling into the 
+                // delegate
+                var lambda = Expression.Lambda<Func<object?, HttpContext, object?, ValueTask>>(body, TargetArg, HttpContextParameter, DeserializedBodyArg);
+                var invoker = lambda.Compile();
+
+                requestDelegate = async (target, httpContext) =>
+                {
+                    var bodyValue = await httpContext.Request.ReadFromJsonAsync(bodyType!);
+
+                    await invoker(target, httpContext, bodyValue);
+                };
+            }
+            else if (needForm)
+            {
+                var lambda = Expression.Lambda<Func<object?, HttpContext, ValueTask>>(body, TargetArg, HttpContextParameter);
+                var invoker = lambda.Compile();
+
+                requestDelegate = async (target, httpContext) =>
+                {
+                    // Generating async code would just be insane so if the method needs the form populate it here
+                    // so the within the method it's cached
+                    await httpContext.Request.ReadFormAsync();
+
+                    await invoker(target, httpContext);
+                };
+            }
+            else
+            {
+                var lambda = Expression.Lambda<Func<object?, HttpContext, ValueTask>>(body, TargetArg, HttpContextParameter);
+                var invoker = lambda.Compile();
+
+                requestDelegate = invoker;
+            }
+
+            return httpContext =>
+            {
+                return requestDelegate(action.Target, httpContext).AsTask();
+            };
+        }
+
+        private static Expression BindParamenter(Expression sourceExpression, ParameterInfo parameter, string name)
```

Comment:
This probably needs to be optimized (eventually) since we're looking at the startup code path now.